### PR TITLE
Netdev

### PR DESCRIPTION
**Pull request by scottfeldman**
(I'm closing the original PR https://github.com/tinygo-org/drivers/pull/523 and submitting this one as a replacement).

This PR adds a network device driver model called netdev. There is a companion PR for TinyGo to update the "net" package here https://github.com/tinygo-org/tinygo/pull/3614. This PR is for the netdev interface definitions, netdev drivers (wifinina, rtl8720dn, espat) and network examples.

See netdev documentation here: README-net.md.

Testing results are [here](https://docs.google.com/spreadsheets/d/e/2PACX-1vT0cCjBvwXf9HJf6aJV2Sw198F2ief02gmbMV0sQocKT4y4RpfKv3dh6Jyew8lQW64FouZ8GwA2yjxI/pubhtml?gid=1013173032&single=true).

### CODE REVIEW

**Comment by scottfeldman** [comment]
Build fails because this PR is dependent on https://github.com/tinygo-org/tinygo/pull/3614.

**Comment by scottfeldman** [comment]
Build fails because this PR is dependent on tinygo-org/tinygo#3614.

**Comment by deadprogram** [comment]
I just tested this branch with the code from https://github.com/tinygo-org/tinygo/pull/3704 on my pyportal 

These examples all worked perfectly, I did not have a chance to test any others yet.

### http-get

```
$ tinygo flash -target pyportal -ldflags="-X main.ssid=AAAA -X main.pass=BBBB" -monitor ./examples/net/http-get/                                                        
Connected to /dev/ttyACM0. Press Ctrl-C to exit.                                                                                                                             
                                                                                                                                                                             
Tinygo ESP32 Wifi network device driver (WiFiNINA)                                                                                                                           
                                                                                                                                                                             
Driver version           : 0.0.1                                                                                                                                             
ESP32 firmware version   : 1.7.4                                                                                                                                             
MAC address              : aa:bb:cc:dd:ee:ff                                                                                                                                 
                                                                                                                                                                             
Connecting to Wifi SSID 'AAAA'...CONNECTED                                                                                                                                   
                                                                                                                                                                             
DHCP-assigned IP         : 192.168.1.223                                                                                                                                     
DHCP-assigned subnet     : 255.255.255.0                                                                                                                                     
DHCP-assigned gateway    : 192.168.1.1                                                                                                                                       
                                                                                                                                                                             
Getting https://httpbin.org/get?name=John+Doe&occupation=gardener                                                                                                            
                                                                                                                                                                             
HTTP/1.1 200 OK                                                                                                                                                              
Server: gunicorn/19.9.0                                                                                                                                                      
Access-Control-Allow-Origin: *                                                                                                                                               
Access-Control-Allow-Credentials: true                                                                                                                                       
Date: Thu, 04 May 2023 20:31:59 GMT                                                                                                                                          
Content-Type: application/json                                                                                                                                               
Content-Length: 332                                                                                                                                                          
Connection: keep-alive                                                                                                                                                       
                                                                                                                                                                             
{                                                                                                                                                                            
  "args": {                                                                                                                                                                  
    "name": "John Doe",                                                                                                                                                      
    "occupation": "gardener"                                                                                                                                                 
  },                                                                                                                                                                         
  "headers": {                                                                                                                                                               
    "Host": "httpbin.org",                                                                                                                                                   
    "User-Agent": "Go-http-client/1.1",
```

### ntpclient

```
$ tinygo flash -target pyportal -ldflags="-X main.ssid=AAAA -X main.pass=BBBB" -monitor ./examples/net/ntpclient/                                                       
Connected to /dev/ttyACM0. Press Ctrl-C to exit.                                                                                                                             
                                                                                                                                                                             
Tinygo ESP32 Wifi network device driver (WiFiNINA)                                                                                                                           
                                                                                                                                                                             
Driver version           : 0.0.1                                                                                                                                             
ESP32 firmware version   : 1.7.4                                                                                                                                             
MAC address              : aa:bb:cc:dd:ee:ff                                                                                                                                   
                                                                                                                                                                             
Connecting to Wifi SSID 'AAAA'...CONNECTED                                                                                                                                   
                                                                                                                                                                             
DHCP-assigned IP         : 192.168.1.223                                                                                                                                     
DHCP-assigned subnet     : 255.255.255.0                                                                                                                                     
DHCP-assigned gateway    : 192.168.1.1                                                                                                                                       
                                                                                                                                                                             
Requesting NTP time...                                                                                                                                                       
NTP time: 2023-05-04 20:30:16 +0000 UTC                                                                                                                                      
                                                                                                                                                                             
Disconnected from Wifi SSID 'AAAA'
```

### mqttclient

```
$ tinygo flash -target pyportal -ldflags="-X main.ssid=AAAA -X main.pass=BBBB" -monitor ./examples/net/mqttclient/                                                      
Connected to /dev/ttyACM0. Press Ctrl-C to exit.                                                                                                                             
                                                                                                                                                                             
Tinygo ESP32 Wifi network device driver (WiFiNINA)                                                                                                                           
                                                                                                                                                                             
Driver version           : 0.0.1                                                                                                                                             
ESP32 firmware version   : 1.7.4                                                                                                                                             
MAC address              : aa:bb:cc:dd:ee:ff                                                                                                                                 
                                                                                                                                                                             
Connecting to Wifi SSID 'AAAA'...FAILED (timed out)                                                                                                                          
Connecting to Wifi SSID 'AAAA'...CONNECTED                                                                                                                                   
                                                                                                                                                                             
DHCP-assigned IP         : 192.168.1.223                                                                                                                                     
DHCP-assigned subnet     : 255.255.255.0                                                                                                                                     
DHCP-assigned gateway    : 192.168.1.1                                                                                                                                       

ClientId: tinygo-client-RPUHOVIRML
Connecting to MQTT broker at tcp://test.mosquitto.org:1883
Connected
Subscribed to topic cpu/freq
Message 120.00Mhz received on topic cpu/freq
Message 120.00Mhz received on topic cpu/freq
Message 120.00Mhz received on topic cpu/freq
Message 120.00Mhz received on topic cpu/freq
Message 120.00Mhz received on topic cpu/freq
```

### webserver 

```
$ tinygo flash -target pyportal -ldflags="-X main.ssid=AAAA -X main.pass=BBBB" -stack-size=4KB -monitor ./examples/net/webserver/

Connected to /dev/ttyACM0. Press Ctrl-C to exit.

Tinygo ESP32 Wifi network device driver (WiFiNINA)

Driver version           : 0.0.1
ESP32 firmware version   : 1.7.4
MAC address              : aa:bb:cc:dd:ee:ff

Connecting to Wifi SSID 'AAAA'...FAILED (timed out)
Connecting to Wifi SSID 'AAAA'...CONNECTED

DHCP-assigned IP         : 192.168.1.223
DHCP-assigned subnet     : 255.255.255.0
DHCP-assigned gateway    : 192.168.1.1
```

**Comment by deadprogram** [comment]
cc @sago35 @bgould @conejoninja @soypat

**Comment by deadprogram** [comment]
This branch  here https://github.com/tinygo-org/drivers/tree/scottfeldman-netdev3 resolvs the merge conflicts in your PR @scottfeldman against the latest `dev` branch.

I would like to see this PR merged along with https://github.com/tinygo-org/tinygo/pull/3704 for the next TinyGo release.

File: examples/net/mqttclient/natiu/main.go [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+
+	// Publish on topic
+	pubFlags, _ := mqtt.NewPublishFlags(mqtt.QoS0, false, false)
+	pubVar := mqtt.VariablesPublish{
+		TopicName: []byte(topic),
+	}
+
+	for i := 0; i < 10; i++ {
+		if !client.IsConnected() {
+			log.Fatal("client disconnected")
```
Comment (soypat) [inline examples/net/mqttclient/natiu/main.go]: The `Err()` method on `mqtt.Client` would return the reason for disconnection, maybe that can be added

File: examples/net/mqttclient/natiu/main.go [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+		err = client.PublishPayload(pubFlags, pubVar, []byte(payload))
+		if err != nil {
+			log.Fatal("error transmitting message: ", err)
+		}
+
+		time.Sleep(time.Second)
+
+		err = client.HandleNext()
+		if err != nil {
+			log.Fatal("handle next: ", err)
```
Comment (soypat) [inline examples/net/mqttclient/natiu/main.go:115]: Maybe this is not an ideal use of `HandleNext` as this could block during a `conn.Read()` call until a packet is received. This could potentially deadlock if the `net.Conn` has no timeout set.

Ideally `HandleNext()` is called in its own goroutine. I understand this is limiting for certain microcontrollers like AVR architecture... but those uCs would not have a TCP stack anyways

File: examples/net/mqttclient/natiu/main.go [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+
+	mqtt "github.com/soypat/natiu-mqtt"
+)
+
+var (
+	ssid   string
+	pass   string
+	broker string = "test.mosquitto.org:1883"
+	topic  string = "cpu/freq"
+)
```
Comment (soypat) [inline examples/net/mqttclient/natiu/main.go:32]: What are thoughts on defining `netdev` here as an interface and setting it in `init` functions in their respective board-specific files?

File: examples/net/mqttclient/natiu/main.go (hunk 1 above)
Comment (scottfeldman) [reply examples/net/mqttclient/natiu/main.go]: Fixing with:

```
-                       log.Fatal("client disconnected")
+                       log.Fatal("client disconnected: ", client.Err())
```

File: examples/net/mqttclient/natiu/main.go (hunk 2 above)
Comment (scottfeldman) [reply examples/net/mqttclient/natiu/main.go:115]: We can add a read deadline on the transport:

```
+               conn.SetReadDeadline(time.Now().Add(10*time.Second))
                err = client.HandleNext()
                if err != nil {
                        log.Fatal("handle next: ", err)
```

I'm not sure what timeout value to use.  1 second is good enough for my testing, but I see 10secs used for timeout in earlier contexts, so using 10sec.

For testing, I tried setting the deadline to 1 Millisecond.  I get the expected timeout i/o failure:

```
ClientId: tinygo-client-GMWJGSAPGA
Connecting to MQTT broker at test.mosquitto.org:1883
Subscribed to topic cpu/freq
1970/01/01 00:00:14 handle next: read tcp <nil>->91.121.93.94:1883: i/o timeout
```

Does this solution look good to you?

File: examples/net/mqttclient/natiu/main.go (hunk 3 above)
Comment (scottfeldman) [reply examples/net/mqttclient/natiu/main.go:32]: Not sure I follow the question.  Can you work up an example?

File: examples/net/mqttclient/natiu/main.go (hunk 2 above)
Comment (soypat) [reply examples/net/mqttclient/natiu/main.go:115]: Hmmm- as a simple example I think it's fine as is. A better alternative would be to leave a link to more complete examples at the repository for natiu: https://github.com/soypat/natiu-mqtt

File: examples/net/mqttclient/natiu/main.go (hunk 3 above)
Comment (soypat) [reply examples/net/mqttclient/natiu/main.go:32]: changing main.go to:
```go
var (
	ssid   string
	pass   string
	broker string = "test.mosquitto.org:1883"
	topic  string = "cpu/freq"
	netdev netdev.Netdev // Do we have this interface defined anywhere in drivers? 
)
```

Putting `netdev` declaration inside `init` function as an assignment (espat.go as an example)
```go
func init() {
    netdev = espat.New(&cfg)
}
```

The above code depends on having the Netdev interface defined somwhere.

Comment (scottfeldman) [reply examples/net/mqttclient/natiu/main.go:32]: I tried this:

```
diff --git a/examples/net/tcpclient/main.go b/examples/net/tcpclient/main.go
index 76fd562..63b3d92 100644
--- a/examples/net/tcpclient/main.go
+++ b/examples/net/tcpclient/main.go
@@ -14,12 +14,15 @@ import (
        "machine"
        "net"
        "time"
+
+       "tinygo.org/x/drivers"
 )
 
 var (
        ssid string
        pass string
        addr string = "10.0.0.100:8080"
+       netdev drivers.netdever
 )
 
 var buf = &bytes.Buffer{}
diff --git a/examples/net/tcpclient/rtl8720dn.go b/examples/net/tcpclient/rtl8720dn.go
index 94d5277..fe6b823 100644
--- a/examples/net/tcpclient/rtl8720dn.go
+++ b/examples/net/tcpclient/rtl8720dn.go
@@ -26,4 +26,6 @@ var cfg = rtl8720dn.Config{
        WatchdogTimeout: time.Duration(20 * time.Second),
 }
 
-var netdev = rtl8720dn.New(&cfg)
+func init() {
+       netdev = rtl8720dn.New(&cfg)
+}
```

But it's a no-go because netdever isn't exported by drivers package:

```
+ tinygo flash -monitor -target wioterminal -stack-size=4KB -size short -ldflags '-X "main.ssid=test" -X "main.pass=testtest"' /home/sfeldma/work/tinygo-drivers/examples/net/tcpclient/
# tinygo.org/x/drivers/examples/net/tcpclient
examples/net/tcpclient/main.go:25:17: netdever not exported by package drivers
examples/net/tcpclient/main.go:34:19: netdev.NetConnect undefined (type drivers.netdever has no field or method NetConnect)
```

**Comment by soypat** [comment]
@scottfeldman Heads up: new version v0.5.1 of natiu-mqtt is out with huge bugfixes, new features and minor API changes.

**Comment by scottfeldman** [comment]
@soypat Thanks for the heads-up.  I pulled and tested v0.5.1 natiu-mqqt against the examples/net/mqttclient/natiu.  No issues, looks good.

**Comment by soypat** [comment]
Scott, could we hold off on adding the netlink changes until we'e used the interface with the cyw43439? It would also help lessen the cognitive load of this PR, which has become quite hard to parse mentally. 

Im actually not sure if the netlink interface is final too...

**Comment by scottfeldman** [comment]
The only changes in netlink in this PR is the addition of two methods to the interface for Send/Recv Eth pkts.  And I have high confidence that those are the right additions.  I'd like to see this PR reviewed and merged to drivers:dev sooner than later.  It's been here since Feb 15th.  Holding it off until cyw43439 is done is the tail wagging the dog.

**Comment by scottfeldman** [comment]
I'll back out the netlink changes and push to this PR, tomorrow.

**Comment by scottfeldman** [comment]
This tinygo-drivers PR must be coordinated with tinygo and tinygo-net PRs:

```
tinygo-org/tinygo:dev   <-   tinygo-org/tinygo:net-submodule-netdev3
tinygo-org/net:dev      <-   scottfeldman/tinygo-net:netdev3           (PR#2)
tinygo-org/drivers:dev  <-   scottfeldman/tinygo-drivers:netdev3       (this PR)
```

**Comment by deadprogram** [comment]
@scottfeldman can you please test this branch against the latest updates in the https://github.com/tinygo-org/tinygo/pull/3704 to make sure things are still working?

**Comment by deadprogram** [comment]
> tinygo-org/net:dev      <-   scottfeldman/tinygo-net:netdev3           (PR#2)

This is now done. :sparkle:

**Comment by scottfeldman** [comment]
@deadprogram, thanks for house keeping.

Ok, everything looks good except for one minor issue.  I tested this PR against the latest from:

```
tinygo-org/tinygo:net-submodule-netdev3
tinygo-org/net:dev
```

example/net/* tests working on wifinina & rtl8720dn devices.

The minor issue is net:dev is missing go.mod, go.sum files.  You'll hit this when running make on tinygo:net-submodule-netdev3.  (After submodule --init).

**Comment by deadprogram** [comment]
@scottfeldman the `net` repo does not need its own go.mod file, as it is only meant to be used as a submodule from the main `tinygo` repo. So I am not exactly sure what problem you are referring to?

**Comment by scottfeldman** [comment]
@deadprogram oh, ok.  I'm using go workspaces, so maybe I need to remove it from my go.work file?  In any case, no problem leaving go.mod out of the `net` repo.

**Comment by deadprogram** [comment]
@scottfeldman please see #613
